### PR TITLE
Feature watermarking update

### PR DESCRIPTION
**Pull request by shiveshr**
**Change log description**  
Merging master into feature-watermarking.
Commits being merged:
1. Issue 3669: Update default TLS material 
2. Issue 3596: Improved Exception handling for ReplyProcessor.
3. Issue 3762: Reduced frequency of logging in ZKSegmentContainerMonitor 
4. Issue 3715: replaced the non thread safe map with SimpleMeterRegistry
5. Issue 3736: Reduce log severity for CancellationException inside AppendProcessor
6. Issue 3769: Increase test pod wait timeout to 1 minute 
7. Issue 3479: Increase client throttling limit 
8. Issue 3721: Fix Multicontroller test failure in docker swarm based system test
9. Issue 3538: Fixing unit test StreamSegmentContainerTests.testWriteFence
10. Issue 3655: TOC and section on connection pooling included in Pravega
11. Issue 3772: (SegmentStore) Fixed DurableLogTests.testTruncateWithout
12. Issue 3781: VersionedSerializer support for Compacted Signed Longs 
13. Issue 3575: Enable running of system tests in a Distributed Pravega 
14. Issue 3725: Fixed sporadic NullPointerException in MetricsTest.
15. Issue 3758: Fix typo in controller API call name 
16. Issue 3775: (SegmentStore) Delete attributes from Tier 2 when lazily
17. Issue 3735: Update StreamCut Validation inside a ReaderGroup to hande
18. Issue 3793: Fixed compatibility for deprecated Controller call 
19. Issue 3805: (SegmentStore) BugFix - StorageWriter not shutting down
20. Issue 3463: Ensure completed transactions are not pinged by the client
21. Issue 3724: (SegmentStore) BugFix - Creating an already existing Segment
22. Issue 3811: Fix SpotBugs error on Java 11 build 
23. Issue 3728: Fix for batch client auth issues 
24. Issue 3816: Sporadic failure of BookieFailoverTest in Docker Swarm system tests
25. Issue 3823: Fix incorrect segment name in exception thrown.
26. Issue 3624: BugFix - Segment Seal and reconnect race causes io.netty
27. Issue 3830: Fix flaky test testSegmentSealedFollowedbyConnectionDrop 
28. Issue 3786: Improved the gauge handling to prevent gauge object loss 
29. Issue 3801, 3803: (SegmentStore) Do not keep checkpoint operations' 
30. Issue 3809: Add unit tests for shared/segment classes 
31. Issue 3843 : Add missing config values required for for HDFS system t…
32. Issue 3832: (SegmentStore) Lazily creating the Attribute Index only o…
33. Issue 3851: (SegmentStore) Bugfix for Segment Id assignment never com…
34. Issue 3798: Lower severity of fail to acquire lock exception
35. Issue 3855: (SegmentStore) Bugfix for segment eviction in Read Index (#…
36. Issue 3296: Fix auth failure in internal communications in Standalone…
37. Issue 3835: Improve exception handling to handle Controller is not av…
38. Issue 3785: Fixes bug inside AssertExtensions 
39. Issue 3718: Fixed flaky unit tests in OperationProcessorTests.
40. Issue 3867: RequestSweeper has incorrect exit check in doWhile loop 
41. Issue 3882: ControllerClusterListener test mocks invoke the method on…
42. Issue 3847: Fix issues causing error in extracting client credentials 
43. Issue 3853: (SegmentStore) Bugfix for ConcurrentModificationException…
44. Issue 1478: Make getRemainingBytes asynchronous 
45. Issue 3893: System test logging 
46. Issue 3907: Sporadic build failure in PravegaTablesStreamMetadataTask…
47. Issue 3425: Ensure the Sealed segment writer is removed after all inf…
48. Issue 3889: Improve test code for ControllerServiceImplTest to enable…
49. Issue-3621: use RawClient to simplify SegmentHelper implementation (#…
50. Issue 3904: Fix PravegaTableStreamMetadataStore's garbage collection …
51. Issue 3916: (SegmentStore) Reduce logging verbosity of IllegalContain…
52. Issue 3859: Fix scale_segmentStore script for external access
53. Issue 2330: Delegation token expiry 
54. Issue 3929: (SegmentStore) Cache eviction bug fix.
55. Issue 3933: Calculated byte rate for autoscaling is underestimated 
56. Issue 3937: Resolve JsonWebToken test failures in Java 11 builds 
57. Issue 3921: SegmentHelper should not throw a ConnectionFailedException…
58. Issue 3692: Do not remove counter of parent segment when delete 
59. Issue 3948: Log KeepAlives at Trace level. 
60. Issues 3918, 3919, 3920: Throttle Transactional Writers in System 
61. Issue 3774: Update the default max connection and Netty channel wate
62. Issue 3940: Fix auth resource representation used for stream in list
63. Issue 3440: Fix AnnotatedNoRouteToHostException observed in docker 
64. Issue 3909: AppendBatchSizeTracker bug fix 
65. Issue 3918: (SegmentStore) Table Segment Pre-Caching 
66. Issue 3442: Ensure znode store/streamsInScope is ignored when listing
67. Issue 3975: Unify usage of RawClient and fix issues in existing RawCl…
68. Issues 3958, 3755: Adding sync before zookeeper reads for ZK orderer …
69. Issue 3978: (BookKeeperLog) Fetching appropriate ZNode version. 
70. Issue 3989: Fix license headers 
71. Issue 3686: Reduce stacktrace upon exceptional AsyncSegmentInputStrea…
72. Issue 3999: Soft handle expected exception logs 
73. Issue 4002: Use load balancer external endpoint hostname if IP isn't …
74. Issue 3556: Changed segment splits/merges metrics from counter to gau…
75. Issue 4000: Upgrade protobuf-gradle-plugin 
76. Issue 4011: (SegmentStore) Making Commit Backlog Throttler more aggre…

**Purpose of the change**  
(_e.g._, Fixes #666, Closes #1234)

**What the code does**  
Resolves conflicts with master. Most conflicts are resolved automatically or trivial, but the ones that were explicitly resolved were in following files:
1. EventStreamReaderImpl
2. EventStreamWriterImpl
3. ReaderGroupImpl.java 

And there were several tests where conflict were resolved. 

**How to verify it**  
All unit tests should pass

### CODE REVIEW

File: client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java [hunk 1]
```diff
@@ -268,7 +274,7 @@ private long getUnreadBytesIgnoringRange(Map<Stream, Map<SegmentWithRange, Long>
         for (Entry<Stream, Map<SegmentWithRange, Long>> streamPosition : positions.entrySet()) {
             StreamCut fromStreamCut = new StreamCutImpl(streamPosition.getKey(), dropRange(streamPosition.getValue()));
             StreamCut toStreamCut = computeEndStreamCut(streamPosition.getKey(), endSegments);
-            totalLength += getRemainingBytes(metaFactory, fromStreamCut, toStreamCut);
+            totalLength += Futures.getAndHandleExceptions(getRemainingBytes(metaFactory, fromStreamCut, toStreamCut), RuntimeException::new).longValue();
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java]: @tkaitchuck please focus on this resolution

File: client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java [hunk 2]
```diff
@@ -50,6 +50,8 @@
 import lombok.Synchronized;
 import lombok.extern.slf4j.Slf4j;
 
+import static io.pravega.client.segment.impl.EndOfSegmentException.ErrorType.END_OF_SEGMENT_REACHED;
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java]: @tkaitchuck please review conflict resolutions in this file

**Comment by andreipaduroiu** [comment]
1. Why does the DCO check complain about me?
2. How am I supposed to review this? It has 7,000 LOCs and 264 files ….

**Comment by shiveshr** [comment]
>     Why does the DCO check complain about me?

no idea. I checked all the merges and they have the signature. @fpj said its because of commits from multiple people being included. 

>    How am I supposed to review this? It has 7,000 LOCs and 264 files ….

well, its like the cherry pick builds.. make sure that your commits are fine (ideally it should be because i only have to resolve conflicts in controller and client).

**Pull request merged**